### PR TITLE
Adds Sign in with Apple

### PR DESCRIPTION
**Pull request by drdaz**
This PR adds Sign In With Apple.

I've taken a slightly different approach to this than the other login providers, in that I didn't add a separate framework project for the purpose. Mostly because it didn't seem necessary in this case - unlike Twitter and FB, all the functionality is built in to the OS, and the API is simple. 

I implemented the tests and functionality under ParseUI, since that's the only place it's going to be used.

### CODE REVIEW

**Comment by drdaz** [comment]
Grrr... was hoping I could knock this out before Christmas. The implementation is pretty simple (I've already done it in Swift in my own app). It just doesn't build here, and I really don't know why 😞

**Comment by drdaz** [comment]
The plot thickens. Switching the deployment target of the library and the demo project to 13+ doesn't actually fix this error. So the @available check seems to have nothing to do with the problem.

**Comment by drdaz** [comment]
... and I found a fix. Adding Apple's AuthenticationServices framework to the ParseUIDemo project allows it to build. Which is odd, since that hasn't been necessary where I've done this elsewhere 🤷🏽‍♂️

**Comment by drdaz** [comment]
Yeah, before Christmas was a little optimistic lol.

Also, working with Objective C these days makes my eyes and my soul bleed.

**Comment by drdaz** [comment]
Well, login works from ParseUI. The Apple Sign In support doesn't yet have it's own library like FB and Twitter (and perhaps doing so is overkill?), and there aren't methods to link and unlink users yet. But it's a decent start methinks.

**Comment by drdaz** [comment]
@TomWFox Do you have any idea how the PFResources.h / .m files are generated?

I was just trying to add an Apple icon to ParseUI, but apparently we don't use graphics like that here. I'm also curious as to why we use this approach.

**Comment by TomWFox** [comment]
No idea

**Comment by noobs2ninjas** [comment]
@drdaz seems my pull request is going to fix builds. Still have some failing tests but it should fix the builds and get rid of those errors once i merge it. running my final round of circleci builds now to verify.

**Comment by drdaz** [comment]
@noobs2ninjas Ahh sweet. I was going to look at the Cocoapods build today actually... are you covering that too?

**Comment by noobs2ninjas** [comment]
Yup! Finally finishing all that stuff from before the holidays trying to update to our own fork of Bolts and the mess that comes with it. I canceled your builds for this for the moment on CircleCI. Once I get that merged all you have to do is catch up to master and well push this through.

**Comment by noobs2ninjas** [comment]
Ok so the deployment script is failing only because we need to update the xcbuildtools repo.

**Comment by noobs2ninjas** [comment]
@drdaz just need an approval on xctoolchain pull request.

**Comment by noobs2ninjas** [comment]
@drdaz after we get my pull request finalized and this merged did we want to start working on pushing a release?

**Comment by drdaz** [comment]
@noobs2ninjas Don't wait for this. We can make another release when it's done. It's putting up a bit of a fight.

**Comment by noobs2ninjas** [comment]
@drdaz my pull request should fix the errors you where experiencing earlier. On my very last build now. So, fingers crossed, Ill merge it in, You can catch up to master, and most of the issues you faced will go away.

**Comment by noobs2ninjas** [comment]
Nevermind. You have different x86 errors. Mine came from catalyst linking in a iOS only target.

**Comment by drdaz** [comment]
I suspect the merge is going to be 'fun'; we've probably both been fiddling with the project files.

I had to disable Catalyst for ParseUI too to get this working. Should probably try and re-enable that before releasing this.

So like I said, don't wait for this PR to release if your changes fix the build.

> Nevermind. You have different x86 errors. Mine came from catalyst linking in a iOS only target.

Are you looking at the screenshot I posted in the first comment? If so, those issues are luckily fixed :)

**Comment by noobs2ninjas** [comment]
@drdaz yea. Similar issue, different method sounds like. So yea, my pull request should clear some of those up. There where two main issues. One is that since macOS no longer supports 32 bit, we needed to remove a indicator that specified x86. Since thats now a given. The second one was that the macOS build which was just looking for let say "Bolts" would mix up identically named target dependencies. For instance we have "Bolts" for iOS-Dynamic(Catalyst), "Bolts" for iOS(which is automatically converted to dynamic. and "Bolts" for macOS. So, many of the times it would grab a dependency that wasn't compatible with older versions of macOS or youd get some weird symbol differences.

**Comment by drdaz** [comment]
That sounds similar to an issue we were seeing on LiveQuery? The one where an iOS dependency was being built for a macOS test?

**Comment by drdaz** [comment]
Oh man... all those merge conflicts all of a sudden 😢

**Comment by noobs2ninjas** [comment]
@drdaz sorry man. Haha I had so much I had to change to get all targets building in CircleCI and updated with latest xcode and building with travis. Gotta Say though. This was worth it. 
Build__1805_-_parse-community_Parse-SDK-iOS-OSX_-_Travis_CI

**Comment by drdaz** [comment]
> @drdaz sorry man. Haha I had so much I had to change to get all targets building in CircleCI and updated with latest xcode and building with travis. Gotta Say though. This was worth it.
> Build__1805_-_parse-community_Parse-SDK-iOS-OSX_-_Travis_CI

Looks good! Nicely done 😃

**Comment by noobs2ninjas** [comment]
At some point we really need to clean up the dynamic targets now that pretty much all iOS targets got converted to catalyst? Theres practically 2 of everything. ParseTwitterUtils-iOS and ParseTwitterUtils-Dynamic. Whats the difference? Nothing at all.

**Comment by drdaz** [comment]
That's something I've thought about too. Having both types of build definitely makes some things kinda tiresome.

I *think* the static and dynamic builds are in place so that the library is usable on both macOS (without Catalyst) and iOS-type OS'. I don't have enough (any) experience with Catalyst, so I don't know if we'd be ok dropping the traditional macOS support.

But if we did, we'd be dropping support for all macOS < 10.15, which seems a little harsh just now 😬

**Comment by noobs2ninjas** [comment]
@drdaz one thing I thought I'd mention. I havent ran your code yet but saw mention of the apple icon. Not sure if you knew this but I thought I'd mention it just in case. From what I understand apple is [VERY specific](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/) on what size/color/shape their buttons are. Now, I know that lots of companies are such as Facebook. Granted, Facebook doesnt bother most apps primarily because theres no real way to keep up with it. However Apple reviews EVERY app and will see each and every app that uses apple sign-in.  I could be completely wrong here but I'm pretty sure, unlike the human interface guidelines, this is much more than a simple "best practices".

**Comment by noobs2ninjas** [comment]
So I read in to it a bit further. Bad news. Its non-negotiable. Good news. They have button sub classes that make it easy. So, no images or resources required.

**Comment by drdaz** [comment]
I’ve got an app in the store with an unofficially sourced logo 🤷🏼‍♂️

I know they provide a button, but it will present differently to ParseUI’s other login buttons.

Can you provide a link to where they prohibit using your own button for Sign In with Apple?

>

**Comment by drdaz** [comment]
In the link you provided, it’s suggested to use Apple’s button:

“ To help people set up an account and sign in, it's *best* to use the familiar buttons that Apple provides for Sign in with Apple.”

**Comment by noobs2ninjas** [comment]
@drdaz Nm. I misread this. Forget it. Basically they are saying that you have have to display the apple logo and its corresponding font as legally required within the developer terms and conditions. Basically dont make up your own Apple logo or associate them with like the Android font with material design.

**Comment by drdaz** [comment]
But where do we get the logo? 😃

**Comment by noobs2ninjas** [comment]
@drdaz good point. I dont see anything but the button subclass and the sign in controller etc. hmm.

**Comment by noobs2ninjas** [comment]
If i where apple I'd leave a vector rather than tempt developers to butcher your logo and risk all sorts of rejected app reviews.

**Comment by noobs2ninjas** [comment]
@drdaz seems you are right. While their verbiage on the Human Interface Guidelines makes it seem like its mandatory to use their exact button. [This page ](https://developer.apple.com/design/human-interface-guidelines/sign-in-with-apple/overview/buttons/) is purely tips for display in various forms proves that is not the case. It seems the only real rule is that you cant create your own artistic interpretation of their logo or font. I'd give this page a look over if you haven't already. Seems they are very insistent about a few things based on the way they are displayed and written.

**Comment by noobs2ninjas** [comment]
Thanks for this BTW @drdaz! Added Icons and properly added documentation for functions. Even got tests added. Well done! 

Let's remind each other to add this to docs before our next release.

**Comment by drdaz** [comment]
Thanks!

It's not done yet though, there's a test file that gets included in the cocoapods build and causes it to fail.

**Comment by noobs2ninjas** [comment]
@drdaz want me to take care of it or do you got it. Lol I've spent so much time on builds that it wouldnt take me but a few seconds.

**Comment by noobs2ninjas** [comment]
After looking at this and seeing this error.

 NOTE  | [Parse/Core, Parse/FacebookUtils, Parse/FacebookUtils-tvOS, and more...] xcodebuild:  note: Execution policy exception registration failed and was skipped: Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted" (in target 'Parse-Parse' from project 'Pods')

After looking around it could have to do with build system.  Looking at the error it has a not that says 

```
NOTE  | [Parse/Core, Parse/FacebookUtils, Parse/FacebookUtils-tvOS, and more...] xcodebuild:  note: Using new build system
```

Seems some people are having luck getting past this by passing in `UseModernBuildSystem=NO` with the xcbuild command.

**Comment by noobs2ninjas** [comment]
Also, FYI the SignInWithAppleTest has a team assigned to it in signing and capabilities.

**Comment by drdaz** [comment]
That's not actually the blocking issue for me, but I agree it should be fixed. Not by dropping back to the legacy build system though.

The issue is if I try to use the SDK with Cocoapods in a test app, that a test case file comes with the pod. That's not right :)

> Also, FYI the SignInWithAppleTest target has a team assigned to it in signing and capabilities.

Thanks I'll fix that.

**Comment by noobs2ninjas** [comment]
Also, we no longer have Travis? Just noticed it on this PR.

**Comment by drdaz** [comment]
> Also, we no longer have Travis? Just noticed it on this PR.

That's really strange! It was here after I merged from master 🧐. You can see it in the build log.

**Comment by drdaz** [comment]
I think this is good now. Ready for feedback at least. 

Travis is back too 🤷🏽‍♂️

**Comment by drdaz** [comment]
Just updated the main text of this, so that it actually describes the PR, rather crying for help 😃

**Comment by cbaker6** [comment]
@drdaz thanks for doing this!

One comment about travis and its long build times. I'm sure this is because travis was updated to 11.4. I made a fix to one of Apple's frameworks on GitHub that made the move and was experiencing long build times like here and test cases started failing. My guess is this because travis moved to a Catalina image for 11.4 and above and they use some sort of a virtual machine which still has bugs with Catalina. On the Apple framework, I switched it to, `osx_image: xcode11.3` which still uses a Mojave image and has reasonable build times with Xcode. 

Travis does currently have a `osx_image: xcode11.5` available, but my guess is that 11.4 still has these long build times, 11.5 has them too. It doesn't seem like this framework needs to be on the bleeding edge of Xcode images and can run on 11.3 until the bugs are worked out

**Comment by noobs2ninjas** [comment]
@cbaker6 good observation. I had to add wait to the cocoapods build time because it was timing out. If I remember right, macOS Catalina actually received some much needed attention when it came to virtualization. My guess is that they specifically did this for use with SwiftUI. Whats strange is my local machine has seen an increase in time since upgrading to Xcode 11 and Catalina and I run macOS natively on my desktop(aka Hackintosh). So, I don't think it's just virtualization.

**Comment by noobs2ninjas** [comment]
@drdaz good job on this. I could learn a thing or two from you about writing unit tests.

**Comment by cbaker6** [comment]
@noobs2ninjas I wonder if there are some lingering issues with Catalina not running on native machines? I build and run tests on my MacBook Pro and don't have issues with timeouts or tests failing. Some of the issues I found online travis and Xcode for Catalina were mentioning "extra calculations occurring" which were causing timeouts

**Comment by noobs2ninjas** [comment]
@cbaker Thats a good point. I dont think we ever had this issue prior to updating environments from XCode11.2.1 which uses macOS 10.14.6 to XCode11.4.1 which uses macOS 10.15.4. Good thinking. Although the wait command was working for us, the process seems to be taking a little longer. I have a potential fix though.

I saw a few similar issues on cocoapods git suggested it could be certain versions of ruby gems no longer print by default, others say it's due to the environment. We've been stuck at Cocoapods 1.8.4. I updated to 1.9.3 and the travis environment to 11.6. Hopefully this works it out. Worst case adding to the end of the command  ``--verbose | ruby -e 'ARGF.each{ print "." }' `` it should print everything as it goes keeping output moving.

**Comment by noobs2ninjas** [comment]
@cbaker Sorry, I thought that was why this thing hadn't been merged and closed.

**Comment by drdaz** [comment]
@noobs2ninjas I'd like you to revert this change. Any issues with build time aren't generated from this PR, and there's enough going on in here without trying to solve a different issue at the same time. If you have a working fix for the build time issue, please create a separate PR.

Only reason this isn't merged yet, is because it got away from me.

**Comment by noobs2ninjas** [comment]
done

**Comment by drdaz** [comment]
Thank you 😊🙏🏼

**Comment by drdaz** [comment]
@mtrezza How's your obj-c 🙂? Could you review this?

File: ParseUI/Sign In With Apple/PFAppleUtils.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+API_AVAILABLE(ios(13.0))
+@interface PFAppleLoginManager : NSObject <ASAuthorizationControllerDelegate, ASAuthorizationControllerPresentationContextProviding>
+
+@end
+
+API_AVAILABLE(ios(13.0))
+@interface PFAppleUtils : NSObject
+
++ (BFTask<NSDictionary *> *)logInInBackground;
+//+ (BFTask<NSNumber *> *)linkUserInBackground:(PFUser *)user;
```
Comment (mtrezza) [inline ParseUI/Sign In With Apple/PFAppleUtils.h:28]: Can we remove these comments?

Comment (drdaz) [reply ParseUI/Sign In With Apple/PFAppleUtils.h]: Ahh yes, sure 🙂

**Comment by mtrezza** [comment]
@drdaz Can you help me understand the failing tests? I understand they are not related to this PR, but what are we doing about it?

**Comment by drdaz** [comment]
> @drdaz Can you help me understand the failing tests? I understand they are not related to this PR, but what are we doing about it? 

Sure, I'll try 😊. So we're not actively doing much about the failing tests because we don't believe they're failing due to issues with the implementation. I'm not aware of evidence that the functionality that fails testing actually fails in use. I'm pretty sure the failures date back to when Flo was maintaining the SDK. 

My last attempt at reproducing one of the errors in Xcode consisted of me repeatedly running one of the tests in question for about 30 minutes. And it just kept passing 🤪. It wasn't great.

> And do we currently merge PRs despite the failing tests?

Yes. But I won't approve changes that cause new test failures.

**Comment by mtrezza** [comment]
@drdaz Thanks.

> Yes. But I won't approve changes that cause new test failures.

How do I know whether a test is one of these usual, randomly failing tests, or a "new test failure"?

After all, a PR can pass its newly introduced tests but break an existing functionality and therefore make an existing test fail.

**Comment by drdaz** [comment]
> How do I know whether a test is one of these usual, randomly failing tests, or a "new test failure"?

The CI build history, unfortunately. If it's been failing regularly for a few years, it's one of those. testFetchPin and testDownloadStreamShares (on both Mac- and iOS) are both regulars. As is the OAuth1FlowDialogTests failure in TwitterUtils.

(Unrelated, but we should look at deprecating or reimplementing PFTwitterUtils... it's partially functional at best, and portions of the code assume the old system-level integration we had for a few iOS versions that's long gone) 

> If these are always the same tests that are randomly failing, can we just disable them?

We could! I confess I hadn't quite given up on getting them fixed. I fear if we kill the tests that never again will our attention return to them 🙂. That said, it seems unlikely that we're going to be making changes in the core SDK code again, so one could argue that a big part of what they offer (regression alerts) is unlikely to be used. 

I'm open on this.

**Comment by mtrezza** [comment]
@drdaz I think the danger is that failing tests could be overlooked which increases the risk of releasing a bug. I suggest to disable the tests and maybe open an issue regarding this.

In the JS SDK / Parse Server we do have a similar issue and we are thinking about [disabling](https://github.com/parse-community/parse-server/issues/6758) these tests. Currently these tests fail randomly, which means we just restart the build until they pass. That's also not optimal, but at least we don't merge *despite* failing tests, which I understand we do here in the iOS SDK?

**Comment by drdaz** [comment]
Sure I can get behind that. And yeah we've been merging and releasing with failing tests.

**Comment by cbaker6** [comment]
@drdaz from my understanding of the chain, I thought you had this finished? Are you waiting to merge to your testcase fixes first?

**Comment by drdaz** [comment]
@cbaker6 Yeah I was thinking about rebasing this once the test fixes were merged.

I'm not sure if @TomWFox added branch protection to the project yet though (so we can't merge PRs with failing tests). If not we can potentially merge this first.

It needs review either way. And a documentation change now I think of it.

**Comment by TomWFox** [comment]
@drdaz There is no branch protection to prevent merging with failing tests currently.

If @cbaker6 and/or @mtrezza are happy with this PR we can merge as far as I'm concerned.

**Comment by cbaker6** [comment]
The PR lgtm and he addressed @mtrezza concerns about the failing tests in a different PR

**Comment by drdaz** [comment]
@mtrezza These failing tests should be fixed by #1523, yes 🙂

**Comment by mtrezza** [comment]
Since Apple started to reject apps that fail to comply with "Sign in with Apple" guidelines, I think this would be a good time for a release of the SDK that includes that feature.

**Comment by TomWFox** [comment]
Agreed, I don't think there is anything blocking - anyone feel free to create a version bump & changelog PR - only thing that needs to be done is to check the cocoapods token in travis as it likes to expire if not used frequently enough!

**Comment by TomWFox** [comment]
I've added a new cocoapods trunk token - expires January 3rd (TO DO - I think we could stop expiry by having travis do `pod trunk me` once in a while)

**Comment by drdaz** [comment]
Yeah there's a handful of important fixes in master now compared to release. I'll make a release PR in the next days.

About that token... how do we check that?

**Comment by TomWFox** [comment]
`pod trunk me` but I'm not sure there is anyway for you to check it because the token belongs to my 'account'

**Comment by drdaz** [comment]
So I was just following the instructions on the front page for updating the version number: bundle exec rake package:set_version. I've tried a few interpretations of that text (some with spaces, some with brackets) and they all fail. It also looks like we might have changed the way we handle version numbers - the script replaces $MARKETING_VERSION with a concrete number in several files.

@TomWFox do you know anything about this?

**Comment by gorillatapstudio** [comment]
@drdaz Hi, i implemented my code with sign in with apple. However, from the code, it seems the email and full name are not passed through to the PFUser object. Is there any other way that i can fetch these info in log::didLogin method? Also, how to differentiate if it is logged in from FB or Apple from the above callback method? thanks!

**Comment by drdaz** [comment]
@gorillatapstudio Sign In With Apple only passes the user details on the first login. So when you're testing, those details won't be there after the very first time you authenticate with your test user. Are you certain this is not what you're seeing?

You need to remove your app from Sign In With Apple before it provides those details again.

**Comment by gorillatapstudio** [comment]
@drdaz I just tested it with a brand new app. In 	func log(_ logInController: PFLogInViewController, didLogIn user: PFUser), the user object does not have either email or full name data. From the code in your PR, it seems these info are not set to PFUser (i am not familiar with objc so I could read wrong).

In the appleUtil: - (void)authorizationController:(ASAuthorizationController *)controller didCompleteWithAuthorization:(ASAuthorization *)authorization, it didn't use cred.email so i guess email is not passed along the code.
In the PFLogInViewController: -(void)_loginWithApple(), we have PFUser *user = t.result[@"user"]; [sself _loginDidSucceedWithUser:user]; which seems only used the user object from the appleUtil and ignored the fullname.

I suppose in the appleUtil, we need to do user["fullName"] = cred.fullName; user.email = cred.email so that these info will pass all the way to the	func log(_ logInController: PFLogInViewController, didLogIn user: PFUser) 

Lastly i still don't know how to identify whether it's logged in from Apple sign in or Parse sign up.

Thanks for working on this PR! I am not sure why i couldn't get the email and fullname. Thanks for the help!

**Comment by mtrezza** [comment]
> Lastly i still don't know how to identify whether it's logged in from Apple sign in or Parse sign up.

Please take a look at [isLinkedWithAuthType](http://parseplatform.org/Parse-SDK-iOS-OSX/api/Classes/PFUser.html#/c:objc(cs)PFUser(im)isLinkedWithAuthType:).

**Comment by gorillatapstudio** [comment]
@mtrezza  thanks a lot!!

**Comment by drdaz** [comment]
Just to let you guys know I'm not ghosting; I'll look at this soon.

**Comment by drdaz** [comment]
So open question to anybody who's listening here @parse-community/ios-osx... where do our other login methods like FB set the data @gorillatapstudio wants attached to the user object here? I can't see anywhere in our code where the user object gets updated with this.

**Comment by mtrezza** [comment]
I am not sure whether the Parse user object should be updated with meta data from the Apple ID account automatically.

I understand what the OP is trying to achieve here, but I think whether and which and how user metadata is stored should be up to the developer and a manual task.

For Facebook, that metadata currently has to be requested via the Facebook Graph API and that is also a manual process up to the developer.

Historically, the permissions and process of obtaining this kind sensitive user data from Facebook has changed significantly and frequently over the last years.￼ We can expect the same to happen for obtaining data from iOS APIs. This means any automated meta data processing will likely require frequent Parse SDK updates.

A solution to this would be to pass the retrieved `ASAuthorizationAppleIDCredential` to the Parse callback routine. This way the developer can choose which meta data to use and store.

**Comment by gorillatapstudio** [comment]
In FB, there is a static variable stores fb credentials in access token, so we can get the last name etc from that in didLogIn delegate.

-- David

**Comment by drdaz** [comment]
> In FB, there is a static variable stores fb credentials in access token, so we can get the last name etc from that in didLogIn delegate.
> […](#)

Do you mean in the logInViewController:didLogInUser: method of PFLogInViewController? I'm only seeing a user there.

Do you mean you use the fb credentials to query for the user details?

**Comment by drdaz** [comment]
> A solution to this would be to pass the retrieved `ASAuthorizationAppleIDCredential` to the Parse callback routine. This way the developer can choose which meta data to use and store.

I think this is the direction I was heading. There was a very long break working on this, and I clearly lost track of how far along I was.

This will be a breaking change as far as I can tell.

**Comment by gorillatapstudio** [comment]
`if let accessToken = AccessToken.current {
			let req = GraphRequest(graphPath: "me", parameters: ["fields":"name,first_name, last_name, id, link, birthday, email, gender"], tokenString: accessToken.tokenString, version: nil, httpMethod: HTTPMethod(rawValue: "GET"))
			_ = req.start(completionHandler: { [weak self] (connection, result, error) in
				if(error == nil) {
					if let fbUser = result as? [String:AnyObject]
...`

AccessToken is a static variable from FB SDK. I don't think Apple Sign in has a similar static variable but you can pass the Apple token all the way to the didLogin delegate

**Comment by SOGAPPS** [comment]
> > A solution to this would be to pass the retrieved `ASAuthorizationAppleIDCredential` to the Parse callback routine. This way the developer can choose which meta data to use and store.
> 
> It looks like this is the direction I was heading. There was a very long break working on this, and I clearly lost track of how far along I was.
> 
> This will be a breaking change as far as I can tell.

Hello @drdaz 
Following up on @mtrezza suggestion. 

Do you've a rough ETA on implementation: With regards to when the Apple retrieved `ASAuthorizationAppleIDCredential` will be passed to the Parse callback routine?

**Comment by mtrezza** [comment]
@SOGAPPS see https://github.com/parse-community/Parse-SDK-iOS-OSX/pull/1551

**Comment by drdaz** [comment]
The [PR](https://github.com/parse-community/Parse-SDK-iOS-OSX/pull/1551) just got approved. I hope to get a release pushed during the next week.

**Pull request merged**